File: server/domain/session_business_date.py
"""北京时间下，按实际覆盖时长确定会话在时间书中的归属日。"""

from __future__ import annotations

from datetime import datetime, time, timedelta, timezone


BEIJING_TZ = timezone(timedelta(hours=8))


def session_business_date(start_time: object, end_time: object, fallback: str = "") -> str:
    """Return the Beijing date with the longest covered duration; ties belong to the later date."""
    start = _parse_beijing_time(start_time)
    end = _parse_beijing_time(end_time)
    if start is None or end is None or end <= start:
        return fallback

    coverage: dict[str, float] = {}
    cursor = start
    while cursor < end:
        next_day = datetime.combine(cursor.date() + timedelta(days=1), time.min, BEIJING_TZ)
        segment_end = min(end, next_day)
        date_text = cursor.date().isoformat()
        coverage[date_text] = coverage.get(date_text, 0.0) + (segment_end - cursor).total_seconds()
        cursor = segment_end
    return max(coverage, key=lambda date_text: (coverage[date_text], date_text))


def _parse_beijing_time(value: object) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(str(value).strip().replace("Z", "+00:00"))
    except (TypeError, ValueError):
        return None
    return parsed.replace(tzinfo=BEIJING_TZ) if parsed.tzinfo is None else parsed.astimezone(BEIJING_TZ)

File: core/core/SessionBusinessDate.ts
import { formatBeijingDate, parseBeijingDateTimeMs } from './BeijingTime'

const BEIJING_OFFSET_MS = 8 * 60 * 60 * 1000

const nextBeijingMidnightMs = (value: number): number => {
  const beijing = new Date(value + BEIJING_OFFSET_MS)
  return Date.UTC(beijing.getUTCFullYear(), beijing.getUTCMonth(), beijing.getUTCDate() + 1) - BEIJING_OFFSET_MS
}

/** Return the Beijing date covered for the longest duration; a tie belongs to the later day. */
export function sessionBusinessDate(startTime: string, endTime: string, fallback = ''): string {
  const start = parseBeijingDateTimeMs(startTime)
  const end = parseBeijingDateTimeMs(endTime)
  if (!Number.isFinite(start) || !Number.isFinite(end) || end <= start) return fallback

  const coverage = new Map<string, number>()
  for (let cursor = start; cursor < end;) {
    const segmentEnd = Math.min(end, nextBeijingMidnightMs(cursor))
    const date = formatBeijingDate(new Date(cursor))
    coverage.set(date, (coverage.get(date) ?? 0) + segmentEnd - cursor)
    cursor = segmentEnd
  }
  return [...coverage.keys()].reduce((winner, date) => {
    const winnerDuration = coverage.get(winner) ?? 0
    const duration = coverage.get(date) ?? 0
    return duration > winnerDuration || (duration === winnerDuration && date > winner) ? date : winner
  })
}
